### PR TITLE
ELF & PE compile using current dialects *Do not merge*

### PR DESCRIPTION
**Pull request by shaopeng-gh**
1. PE compile using current dialects. fixed the comments from Eddy, also skip the referenced libraries
2. ELF compile using current dialects.
3. during Elf work found that some case the producer flag data will be in LinkageName, change code to support reading from it.

### CODE REVIEW

File: src/BinSkim.Rules/DwarfRules/BA3012.DwarfCompileUsingCurrentDialects.cs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            reasonForNotAnalyzing = result.Reason;
+            return result.Result;
+        }
+
+        public override void Analyze(BinaryAnalyzerContext context)
+        {
+            IDwarfBinary binary = context.DwarfBinary();
+
+
+
```
Comment (eddynaka) [inline src/BinSkim.Rules/DwarfRules/BA3012.DwarfCompileUsingCurrentDialects.cs:72]: can you clean all this spaces?

File: src/BinSkim.Rules/DwarfRules/BA3012.DwarfCompileUsingCurrentDialects.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            static bool analyze(IDwarfBinary binary)
+            {
+                string dwarfCompilerCommand = binary.GetDwarfCompilerCommand();
+
+                List<string> args = ArgumentSplitter.CommandLineToArgvW(dwarfCompilerCommand);
+
+                return args.Count >= 2 && (!OldDialectsForC.Contains(args[1]) && !OldDialectsForCPlugPlus.Contains(args[1]));
+            }
```
Comment (eddynaka) [inline src/BinSkim.Rules/DwarfRules/BA3012.DwarfCompileUsingCurrentDialects.cs:80]: i would move this section to the place inside if (binary is ElfBinary elf)

File: src/BinSkim.Rules/PERules/BA2027.CompileUsingCurrentDialects.cs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                else
+                {
+                    int cVersionNumber;
+                    if (cVersionNumberString.Equals("latest", StringComparison.OrdinalIgnoreCase))
+                    {
+                        resultType = AnalyzeResultType.Latest;
+                    }
+                    else if (int.TryParse(cVersionNumberString, out cVersionNumber))
+                    {
+                        resultType = cVersionNumber >= 17 ? AnalyzeResultType.Pass : AnalyzeResultType.NotCurrent;
```
Comment (eddynaka) [inline src/BinSkim.Rules/PERules/BA2027.CompileUsingCurrentDialects.cs:126]: >Pass [](http://example.com/codeflow?start=78&length=4)

Change to Current

File: src/BinSkim.Rules/PERules/BA2027.CompileUsingCurrentDialects.cs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                    RuleUtilities.BuildResult(ResultKind.Pass, context, null,
+                    nameof(RuleResources.BA2027_Pass),
+                    context.TargetUri.GetFileName()));
+                return;
+            }
+        }
+
+        private class AnalyzeResult
+        {
+            public ObjectModuleDetails Details { get; set; }
```
Comment (eddynaka) [inline src/BinSkim.Rules/PERules/BA2027.CompileUsingCurrentDialects.cs:187]: order by length. so, this is the third line.

File: src/BinSkim.Rules/PERules/BA2027.CompileUsingCurrentDialects.cs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                else
+                {
+                    int cVersionNumber;
+                    if (cVersionNumberString.Equals("latest", StringComparison.OrdinalIgnoreCase))
+                    {
+                        resultType = AnalyzeResultType.Latest;
+                    }
+                    else if (int.TryParse(cVersionNumberString, out cVersionNumber))
+                    {
+                        resultType = cVersionNumber >= 17 ? AnalyzeResultType.Current : AnalyzeResultType.NotCurrent;
```
Comment (shaopeng-gh) [inline src/BinSkim.Rules/PERules/BA2027.CompileUsingCurrentDialects.cs:128]: >17 [](http://example.com/codeflow?start=55&length=2)

Ah! I just noticed that e.g. C99 with number 99 will be greater then 17. I will re-write this when back on this task.

**Comment by shaopeng-gh** [comment]
Close to clean up prs.

**Pull request closed without merging**